lib/Sema/ConstraintSystem.cpp

```diff
@@ -4850,14 +4911,83 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
   // overload choices.
   fixes.set_subtract(consideredFixes);
 
+  // Aggregate all requirement fixes that belong to the same callee
+  // and attempt to diagnose possible ambiguities.
+  {
+    auto isResultBuilderMethodRef = [&](ASTNode node) {
+      auto *UDE = getAsExpr<UnresolvedDotExpr>(node);
+      if (!(UDE && UDE->isImplicit()))
+        return false;
+
+      auto &ctx = getASTContext();
+      SmallVector<Identifier, 4> builderMethods(
+          {ctx.Id_buildBlock, ctx.Id_buildExpression, ctx.Id_buildPartialBlock,
+           ctx.Id_buildFinalResult});
+
+      return llvm::any_of(builderMethods, [&](const Identifier &methodId) {
+        return UDE->getName().compare(DeclNameRef(methodId)) == 0;
+      });
+    };
+
+    // Aggregates fixes fixes attached to `buildExpression` and `buildBlock`
+    // methods at the particular source location.
+    llvm::MapVector<SourceLoc, SmallVector<FixInContext, 4>>
+        builderMethodRequirementFixes;
+
+    llvm::MapVector<ConstraintLocator *, SmallVector<FixInContext, 4>>
+        perCalleeRequirementFixes;
+
+    for (const auto &entry : fixes) {
+      auto *fix = entry.second;
+      if (!fix->getLocator()->isLastElement<LocatorPathElt::AnyRequirement>())
+        continue;
+
+      auto *calleeLoc = entry.first->getCalleeLocator(fix->getLocator());
+
+      if (isResultBuilderMethodRef(calleeLoc->getAnchor())) {
+        auto *anchor = castToExpr<Expr>(calleeLoc->getAnchor());
+        builderMethodRequirementFixes[anchor->getLoc()].push_back(entry);
+      } else {
+        perCalleeRequirementFixes[calleeLoc].push_back(entry);
+      }
+    }
+
+    SmallVector<SmallVector<FixInContext, 4>, 4> viableGroups;
+    {
+      auto takeAggregateIfViable =
+          [&](const SmallVector<FixInContext, 4> &aggregate) {
+            // Ambiguity only if all of the solutions have a requirement
+            // fix at the given location.
+            if (aggregate.size() == solutions.size())
+              viableGroups.push_back(std::move(aggregate));
```

Comment:
Since aggregate is `const` I think move here results in a copy anyways.  Overload resolution would end picking up const ref overload of `push_back` if I understand correctly.

Reply:
Hm I thought that `SmallVector` has a `const SmallVector<T> &&` overload for constructor but apparently not, although maybe it still makes sense to keep it to show intent because the vector passed there is taken from the map...

Reply:
Does `aggregate` need to be `const`? If is just a non-const ref then it can be moved. 
```cpp
#include <iostream>
#include <vector>
class A {
public:
  A() {}
  A(A&&a) noexcept { std::cout << "Move" << std::endl;  }
  A(const A&a) noexcept { std::cout << "Copy" << std::endl; }
};

int main(int argc, const char * argv[]) {
  std::vector<A> v;
  auto l = [&](const A&a) {
    v.push_back(std::move(a));
  };
  auto l1 = [&](A&a) {
    v.push_back(std::move(a));
  };
  
  A a;
  l(a); // Copy
  l1(a); // Move
  return 0;
}

```

Reply:
Let me try, I though that we kind of stuck with `const` here, but maybe not.

Reply:
Looks like we can indeed drop `const`!

Reply:
Cool :)